src/mscorlib/src/System/Text/StringBuilder.CoreCLR.cs

```diff
@@ -8,11 +8,94 @@ namespace System.Text
 {
     public partial class StringBuilder
     {
-        [MethodImplAttribute(MethodImplOptions.InternalCall)]
-        internal extern unsafe void ReplaceBufferInternal(char* newBuffer, int newLength);
+        private int GetReplaceBufferCapacity(int requiredCapacity)
+        {
+            // This function assumes that required capacity will be less 
+            // than the max capacity of the StringBuilder
+            Diagnostics.Debug.Assert(requiredCapacity <= m_MaxCapacity);
+
+            int newCapacity = Capacity;
+            // Round the current capacity to the nearest multiple of 2
+            // that is larger than the required capacity.
+            if (newCapacity < requiredCapacity)
+            {
+                newCapacity = (requiredCapacity + 1) & ~1;
+            }
+            return newCapacity;
+        }
+
+        internal unsafe void ReplaceBufferInternal(char* newBuffer, int newLength)
+        {
+            if (newLength > m_MaxCapacity)
+                throw new ArgumentOutOfRangeException("capacity", SR.ArgumentOutOfRange_Capacity);
+
+            // Leave space for a NUL character at the end, the previous implementation did this
+            m_ChunkChars = new char[GetReplaceBufferCapacity(newLength + 1)];
+
+            fixed (char* pChunkChars = m_ChunkChars)
```

Comment:
`new Span<char>(pChunkChars, newLength).CopyTo(m_ChunkChars);` ?

Reply:
You mean `new Span<char>(newBuffer, newLength).CopyTo…`. Sure, I suppose that's better than using `fixed`.

Reply:
Yes, that's what I meant.